src/flags/apiVersion.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { Flags, CliUx } from '@oclif/core';
+import { Messages, sfdc } from '@salesforce/core';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/sf-plugins-core', 'messages');
+
+// versions below this are retired
+export const minValidApiVersion = 21;
+// this and all un-retired versions below it are deprecated
+export const maxDeprecated = 30;
+export const maxDeprecatedUrl = 'https://help.salesforce.com/s/articleView?id=000354473&type=1;';
+
+/**
+ * apiVersion for a salesforce org's rest api.
+ * Will validate format and that the api version is still supported.
+ * Will default to the version specified in Config, if it exists (and will provide an override warning)
+ *
+ * CAVEAT: unlike the apiversion flag on sfdxCommand, this does not set the version on the org/connection
+ * We leave this up to the plugins to implement
+ */
+export const apiVersionFlag = Flags.build<string>({
+  parse: async (input: string) => validate(input),
+  default: async () => await getDefaultFromConfig(),
+  description: messages.getMessage('flags.apiVersion.description'),
+});
+
+const getDefaultFromConfig = async (): Promise<string | undefined> => {
+  // (perf) only import ConfigAggregator and cliUX if necessary
+  const { ConfigAggregator } = await import('@salesforce/core');
+  const config = await ConfigAggregator.create();
+  const apiVersionFromConfig = config.getInfo('apiVersion')?.value as string;
+  if (apiVersionFromConfig) {
+    CliUx.ux.warn(messages.getMessage('flags.apiVersion.overrideWarning', [apiVersionFromConfig]));
```

Comment:
We need to automatically suppress this whenever `--json` is used

Reply:
Oh, because it won't automatically go into the json's .warnings array?  

Any good ideas how to do that?

Reply:
Correct. Warnings are captured in the `warn` method of `SfCommand`: https://github.com/salesforcecli/sf-plugins-core/blob/main/src/sfCommand.ts#L63

Looking at the implementation in the [UX class](https://github.com/salesforcecli/command/blob/main/src/ux.ts#L233), it looks like there's a static `warnings` property that holds all of the warnings. That would allow you to add the warning to the class instance but you still have the problem of knowing whether or not `--json` is being used. Maybe you can delay the warnings until after all the flags have been parsed?

Reply:
I used lifecycle to delegate to the command.  We want to have that `onWarning` subscriber at the command level anyway since libraries (STL, SDR, core) also emit warnings using this approach